src/oods_statement.rs

```diff
@@ -0,0 +1,296 @@
+        ];
+
+        // Encoding the 'layout' string to its ASCII byte representation and converting to U256
+        let layout_big = U256::from_big_endian(self.public_input.layout.as_bytes());
+        cairo_aux_input.push(layout_big);
+
+        // Extend with serialized segments
+        let serialized_segments = self.serialize_segments();
+        cairo_aux_input.extend(serialized_segments);
+
+        let z = self.interaction_z;
+        let alpha = self.interaction_alpha;
+
+        let memory_pages_public_input =
+            self.memory_page_public_input(self.public_input.public_memory.clone(), z, alpha);
+
+        // Extend with memory pages public input - assuming this is already a Vec<U256>
+        cairo_aux_input.extend(memory_pages_public_input);
+
+        // Append z and alpha
+        cairo_aux_input.push(z);
+        cairo_aux_input.push(alpha);
+
+        cairo_aux_input
+    }
+
+    /// Serialize memory segments in order
+    fn serialize_segments(&self) -> Vec<U256> {
+        let segment_names = [
+            "program",
+            "execution",
+            "output",
+            "pedersen",
+            "range_check",
+            "ecdsa",
+            "bitwise",
+            "ec_op",
+            "keccak",
+            "poseidon",
+        ];
+
+        let segments = &self.public_input.memory_segments;
+        let mut sorted_segments: Vec<MemorySegment> = Vec::new();
+
+        for name in segment_names.iter() {
+            let segment: Option<&MemorySegment> = segments.get(*name);
+            if let Some(seg) = segment {
+                sorted_segments.push(seg.clone());
+            }
+        }
+
+        assert_eq!(sorted_segments.len(), segments.len());
+
+        let mut result: Vec<U256> = Vec::new();
+        for segment in sorted_segments {
+            result.push(U256::from(segment.begin_addr));
+            result.push(U256::from(segment.stop_ptr));
+        }
+
+        result
+    }
+
+    /// Calculate accumulated product for continuous memory
+    fn calculate_product(
+        prod: U256,
+        z: U256,
+        alpha: U256,
+        memory_address: U256,
+        memory_value: U256,
+        prime: U256,
+    ) -> U256 {
+        let bigint_prod = BigInt::from_str_radix(&prod.to_string(), 10).unwrap();
+        let bigint_alpha = BigInt::from_str_radix(&alpha.to_string(), 10).unwrap();
+        let bigint_z = BigInt::from_str_radix(&z.to_string(), 10).unwrap();
+        let bigint_memory_value = BigInt::from_str_radix(&memory_value.to_string(), 10).unwrap();
+        let bigint_memory_address =
+            BigInt::from_str_radix(&memory_address.to_string(), 10).unwrap();
+        let bigint_prime = BigInt::from_str_radix(&prime.to_string(), 10).unwrap();
+
+        let multiply =
+            bigint_prod * (bigint_z - (bigint_memory_address + bigint_alpha * bigint_memory_value));
+        let mod_multiply = multiply.modpow(&BigInt::one(), &bigint_prime);
```

Comment:
maybe better to use arkworks ff library for these kind of operations

Reply:
IIUC, it requires some efforts to implement the `FpParameters` for the default prime as modulus. I created [backlog](https://github.com/zksecurity/stark-evm-adapter/issues/17) for this.